src/predict.rs

```diff
@@ -113,137 +113,32 @@ pub trait Dim {
   const H: usize;
 }
 
-pub struct Block4x4;
-
-impl Dim for Block4x4 {
-  const W: usize = 4;
-  const H: usize = 4;
-}
-
-pub struct Block8x8;
-
-impl Dim for Block8x8 {
-  const W: usize = 8;
-  const H: usize = 8;
-}
-
-pub struct Block16x16;
-
-impl Dim for Block16x16 {
-  const W: usize = 16;
-  const H: usize = 16;
-}
-
-pub struct Block32x32;
-
-impl Dim for Block32x32 {
-  const W: usize = 32;
-  const H: usize = 32;
-}
-
-pub struct Block64x64;
-
-impl Dim for Block64x64 {
-  const W: usize = 64;
-  const H: usize = 64;
-}
-
-pub struct Block4x8;
-
-impl Dim for Block4x8 {
-  const W: usize = 4;
-  const H: usize = 8;
-}
-
-pub struct Block8x4;
-
-impl Dim for Block8x4 {
-  const W: usize = 8;
-  const H: usize = 4;
-}
-
-pub struct Block8x16;
-
-impl Dim for Block8x16 {
-  const W: usize = 8;
-  const H: usize = 16;
-}
-
-pub struct Block16x8;
-
-impl Dim for Block16x8 {
-  const W: usize = 16;
-  const H: usize = 8;
-}
-
-pub struct Block16x32;
-
-impl Dim for Block16x32 {
-  const W: usize = 16;
-  const H: usize = 32;
-}
-
-pub struct Block32x16;
-
-impl Dim for Block32x16 {
-  const W: usize = 32;
-  const H: usize = 16;
-}
-
-pub struct Block32x64;
-
-impl Dim for Block32x64 {
-  const W: usize = 32;
-  const H: usize = 64;
-}
-pub struct Block64x32;
-
-impl Dim for Block64x32 {
-  const W: usize = 64;
-  const H: usize = 32;
-}
-
-pub struct Block4x16;
-
-impl Dim for Block4x16 {
-  const W: usize = 4;
-  const H: usize = 16;
-}
-
-pub struct Block16x4;
-
-impl Dim for Block16x4 {
-  const W: usize = 16;
-  const H: usize = 4;
-}
-
-pub struct Block8x32;
-
-impl Dim for Block8x32 {
-  const W: usize = 8;
-  const H: usize = 32;
-}
-
-pub struct Block32x8;
-
-impl Dim for Block32x8 {
-  const W: usize = 32;
-  const H: usize = 8;
+macro_rules! block_dimension {
+  ($W:expr, $H:expr) => {
+    paste::item! {
+      pub struct [<Block $W x $H>];
+
+      impl Dim for [<Block $W x $H>] {
+        const W: usize = $W;
+        const H: usize = $H;
+      }
+    }
+  };
 }
 
-pub struct Block16x64;
-
-impl Dim for Block16x64 {
-  const W: usize = 16;
-  const H: usize = 64;
+macro_rules! blocks_dimension {
```

Comment:
Thank you. This is indeed much better!
So, I understand that the rustc itself copies N of "block_dimension! { $W, $H }" here, correct?
Then, I wonder which syntax actually dose that N copying, i.e. iterating?  "$(" .. ")"?

Reply:
The macros have the concept of [repetition](https://doc.rust-lang.org/stable/rust-by-example/macros/repeat.html), you see in `$( ... )*` below how I'm using it. In general you express with `$(some pattern),+` or `$(some pattern),*` in the macro argument position the repeating sequence and then you have it repeated by putting in the body `$(` + `something you want to repeat`  +  `)*`.